pkg/pool-weighted/contracts/lbp/LiquidityBootstrappingPool.sol

```diff
@@ -240,79 +278,389 @@ contract LiquidityBootstrappingPool is BaseWeightedPool, ReentrancyGuard {
+    }
+
+    function _exitExactBPTInForTokenOut(
+        uint256[] memory balances,
+        uint256[] memory normalizedWeights,
+        uint256 totalSupply,
+        bytes memory userData
+    ) private view returns (uint256, uint256[] memory) {
+        (uint256 bptAmountIn, uint256 tokenIndex) = userData.exactBptInForTokenOut();
+        // Note that there is no minimum amountOut parameter: this is handled by `IVault.exitPool`.
+
+        _require(tokenIndex < balances.length, Errors.OUT_OF_BOUNDS);
+
+        uint256 amountOut = WeightedMath._calcTokenOutGivenExactBptIn(
+            balances[tokenIndex],
+            normalizedWeights[tokenIndex],
+            bptAmountIn,
+            totalSupply,
+            getSwapFeePercentage()
+        );
+
+        // This is an exceptional situation in which the fee is charged on a token out instead of a token in.
+        // We exit in a single token, so we initialize amountsOut with zeros
+        uint256[] memory amountsOut = new uint256[](balances.length);
+        // And then assign the result to the selected token
+        amountsOut[tokenIndex] = amountOut;
+
+        return (bptAmountIn, amountsOut);
+    }
+
+    function _exitExactBPTInForTokensOut(
+        uint256[] memory balances,
+        uint256 totalSupply,
+        bytes memory userData
+    ) private pure returns (uint256, uint256[] memory) {
+        uint256 bptAmountIn = userData.exactBptInForTokensOut();
+        // Note that there is no minimum amountOut parameter: this is handled by `IVault.exitPool`.
+
+        uint256[] memory amountsOut = WeightedMath._calcTokensOutGivenExactBptIn(balances, bptAmountIn, totalSupply);
+        return (bptAmountIn, amountsOut);
+    }
+
+    function _exitBPTInForExactTokensOut(
+        uint256[] memory balances,
+        uint256[] memory normalizedWeights,
+        uint256[] memory scalingFactors,
+        uint256 totalSupply,
+        bytes memory userData
+    ) private view returns (uint256, uint256[] memory) {
+        (uint256[] memory amountsOut, uint256 maxBPTAmountIn) = userData.bptInForExactTokensOut();
+        InputHelpers.ensureInputLengthMatch(amountsOut.length, balances.length);
+        _upscaleArray(amountsOut, scalingFactors);
+
+        // This is an exceptional situation in which the fee is charged on a token out instead of a token in.
+        uint256 bptAmountIn = WeightedMath._calcBptInGivenExactTokensOut(
+            balances,
+            normalizedWeights,
+            amountsOut,
+            totalSupply,
+            getSwapFeePercentage()
+        );
+        _require(bptAmountIn <= maxBPTAmountIn, Errors.BPT_IN_MAX_AMOUNT);
+
+        return (bptAmountIn, amountsOut);
     }
 
     // Swap overrides - revert unless swaps are enabled
 
+    function _onSwapGeneral(
+        SwapRequest memory,
+        uint256[] memory,
+        uint256,
+        uint256
+    ) internal virtual override returns (uint256) {
+        _revert(Errors.UNHANDLED_BY_LBP);
```

Comment:
```suggestion
        _revert(Errors.UNIMPLEMENTED);
```

I don't think we need to add a new error type for this.

Reply:
It was patterned off Linear Pool and others, that used "UNHANDLED_BY" for things like joins and exits - and unsupported swap types. Should we replace all the "UNHANDLED" ones with UNIMPLEMENTED then?

Reply:
Are they actually used? I can't find any usage of `UNHANDLED_BY_PHANTOM_POOL`, `UNHANDLED_BY_MANAGED_POOL`.

I'm not a fan of these message generally as they're all reporting the same underlying issue, "You're asking me to perform an action for which I don't have code to handle".

Reply:
They *were* used, but it looks like now they're only in Linear. Ok; I'm not a big fan either - especially having multiple "versions" of these kinds of messages. UNHANDLED also has the sense of "I should be able to do this, but can't" - while UNIMPLEMENTED seems more intentional.

I'll change this one then, and make a new PR to change Linear - and we'll not introduce any further UNHANDLED ones.